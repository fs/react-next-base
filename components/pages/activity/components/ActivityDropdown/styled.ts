import styled, { css, DefaultTheme, FlattenSimpleInterpolation } from 'styled-components';

export const Wrapper = styled.div<{
  customStyles?: (theme: DefaultTheme) => FlattenSimpleInterpolation | string;
}>(
  ({ theme, customStyles }) => css`
    display: flex;
    justify-content: center;

    ${customStyles && customStyles(theme)}
  `,
);

export const StyledLabel = styled.label`
  display: flex;
  flex-direction: column;
  color: ${({ theme }) => theme.colors.darkGrey};
`;

export const StyledSelect = styled.select(
  ({ theme: { up, breakpoints } }) => css`
    margin-top: 0.5rem;
    padding: 0.25rem;
    font-size: 1rem;

    ${up(breakpoints.sm)} {
      font-size: 0.8rem;
    }
  `,
);
