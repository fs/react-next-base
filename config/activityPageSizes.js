export default [5, 10, 25, 50];
