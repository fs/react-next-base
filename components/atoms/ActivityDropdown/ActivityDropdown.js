import React from 'react';
import styled, { css } from 'styled-components';

const Wrapper = styled.div(
  ({ theme, customStyles }) => css`
    display: flex;
    justify-content: center;

    ${customStyles && customStyles(theme)}
  `,
);

const StyledLabel = styled.label`
  display: flex;
  flex-direction: column;
  color: ${({ theme }) => theme.colors.darkGrey};
`;

const StyledSelect = styled.select(
  ({ theme: { up, breakpoints } }) => css`
    margin-top: 0.5rem;
    padding: 0.25rem;
    font-size: 1rem;

    ${up(breakpoints.sm)} {
      font-size: 0.8rem;
    }
  `,
);

const ActivityDropdown = ({
  label,
  selectedValue,
  values,
  hasEmptyOption,
  emptyOptionLabel,
  onChange,
  disabled,
  customStyles,
  testId,
}) => {
  return (
    <Wrapper customStyles={customStyles} data-testid={testId}>
      <StyledLabel htmlFor="activity-dropdown">
        {label}
        <StyledSelect
          id="activity-dropdown"
          data-cy={testId}
          value={selectedValue}
          onChange={onChange}
          disabled={disabled}
        >
          {hasEmptyOption && <option value="">{emptyOptionLabel || '--Please choose an option--'}</option>}
          {values.map(({ value, name }) => (
            <option key={value} value={value}>
              {name}
            </option>
          ))}
        </StyledSelect>
      </StyledLabel>
    </Wrapper>
  );
};

export default ActivityDropdown;
