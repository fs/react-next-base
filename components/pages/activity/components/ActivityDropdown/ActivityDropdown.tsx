import React from 'react';
import { DefaultTheme, FlattenSimpleInterpolation } from 'styled-components';

import Test from 'types/testType';
import { Wrapper, StyledLabel, StyledSelect } from './styled';

type Option = {
  value: string | number;
  name: string | number;
};

type Props = Test & {
  label: string;
  selectedValue?: Option['value'];
  values: Option[];
  hasEmptyOption?: boolean;
  emptyOptionLabel?: string;
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  disabled?: boolean;
  customStyles?: (theme: DefaultTheme) => FlattenSimpleInterpolation;
};

const ActivityDropdown = ({
  label,
  selectedValue,
  values,
  hasEmptyOption,
  emptyOptionLabel,
  onChange,
  disabled,
  customStyles,
  testId,
}: Props): JSX.Element => {
  return (
    <Wrapper customStyles={customStyles} data-testid={testId}>
      <StyledLabel htmlFor="activity-dropdown">
        {label}
        <StyledSelect
          id="activity-dropdown"
          data-cy={testId}
          value={selectedValue}
          onChange={onChange}
          disabled={disabled}
        >
          {hasEmptyOption && <option value="">{emptyOptionLabel || '--Please choose an option--'}</option>}
          {values.map(({ value, name }) => (
            <option key={value} value={value}>
              {name}
            </option>
          ))}
        </StyledSelect>
      </StyledLabel>
    </Wrapper>
  );
};

export default ActivityDropdown;
