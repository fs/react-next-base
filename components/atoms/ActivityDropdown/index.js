import ActivityDropdown from './ActivityDropdown';

export default ActivityDropdown;
