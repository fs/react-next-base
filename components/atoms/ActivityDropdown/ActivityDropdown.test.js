import React from 'react';
import { render, screen } from '@testing-library/react';
import 'jest-styled-components';

import renderWithTheme from '__tests__/helpers/renderWithTheme';
import activityPageSizes from 'config/activityPageSizes';

import ActivityDropdown from 'components/atoms/ActivityDropdown';

describe('ActivityDropdown', () => {
  const mockLabel = "ActivityDropdown's label";
  const mockValues = activityPageSizes.map(item => ({ value: item, name: item }));
  const mockTestId = 'test-activity-dropdown';

  test('should render correctly', () => {
    // Arrange

    // Act
    render(renderWithTheme(<ActivityDropdown label={mockLabel} values={mockValues} testId={mockTestId} />));

    const dropdown = screen.getByTestId(mockTestId);

    // Assert
    expect(dropdown).toMatchSnapshot();
  });

  test('should render empty option', () => {
    // Arrange
    const mockEmptyOptionLabel = 'Choose option';
    const expectedValuesCount = activityPageSizes.length + 1;

    // Act
    render(
      renderWithTheme(
        <ActivityDropdown
          label={mockLabel}
          values={mockValues}
          hasEmptyOption
          emptyOptionLabel={mockEmptyOptionLabel}
        />,
      ),
    );

    const emptyOption = screen.getByText(mockEmptyOptionLabel);
    const options = screen.getAllByRole('option');

    // Assert
    expect(emptyOption).toBeDefined();
    expect(options.length).toEqual(expectedValuesCount);
  });
});
