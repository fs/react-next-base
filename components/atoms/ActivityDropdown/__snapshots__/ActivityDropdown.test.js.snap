// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityDropdown should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #606c76;
}

.c2 {
  margin-top: 0.5rem;
  padding: 0.25rem;
  font-size: 1rem;
}

@media (min-width:calc(576px + 0.02px)) {
  .c2 {
    font-size: 0.8rem;
  }
}

<div
  class="c0"
  data-testid="test-activity-dropdown"
>
  <label
    class="c1"
    for="test-activity-dropdown-dropdown"
  >
    ActivityDropdown's label
    <select
      class="c2"
      id="test-activity-dropdown-dropdown"
    >
      <option
        value="5"
      >
        5
      </option>
      <option
        value="10"
      >
        10
      </option>
      <option
        value="25"
      >
        25
      </option>
      <option
        value="50"
      >
        50
      </option>
    </select>
  </label>
</div>
`;
